### PR TITLE
Optimisation of gutter logic

### PR DESCRIPTION
**Pull request by steve-holland**
**What kind of change is this? (Bug Fix, Feature...)**
Optimisation.

**What is the current behavior (You can also link to an issue)**
#371  - As per issue raised.

**What is the new behavior this introduces (if any)**
Simplifies output of gutter logic.

**Does this introduce any breaking changes?**
No.

**Does the PR fulfill these requirements?**
- [✔️ ] Tests for the changes have been added
- [ N/A] Docs have been added or updated

**Other Comments**
Docs I'll look at updating at a later date. Just incase the output this generates needs changing first.

### CODE REVIEW

**Comment by steve-holland** [comment]
I'd really appreciate you having a good look around this merge as it does have far-reaching changes, which makes me quite nervous!

As I essentially have to perform part of the calculation `calc` is doing, I wasn't sure about how accurate the rounding of the result needs to be. I've gone for 5 decimal places, but it doesn't have to be. We'd just need to update the associated tests. 

As always, constructive criticism and nitpicking gratefully received :)

File: test/lg-logic.js
Comment (peterramsing) [inline test/lg-logic.js:12]: Could we pull off the trailing zeros within the `calcValue` function? CSS Clean looks to be stripping them out for us in the tests, but without it, we're leaving an optimization on the table. What're five zeroes though? Nothing worth splitting hairs over *but* if we're already doing optimizations it would be cool to take it all the way. **not blocking merge**

**Review by peterramsing — CHANGES_REQUESTED** [review]
I spent some time reading into the spec and couldn't find a specific "rounding" rule that the browsers use for calc. It looks like `postcss-calc` defaults to 5 decimal points of precision, but even that is configurable. I think your default of 5 is the correct choice if LostGrid should get into the precision work *but* I'm thinking that we might want to leave that up to the user. Sorry that this conversation happened after you got to work (sidebar: let's talk about if we want to add a tailboarding process) but I have a strong feeling that unless we can calculate the CSS to a whole number then we should let the user optimize it themselves (using `postcss-calc` or `cssnano` or another tool) *or* leave it to the browser to decide. I think that keeps LostGrid focused on the grid and will help users find bugs caused by rounding if LostGrid isn't doing that for them. 

If we do want to get into this I propose we offer a similar parameter, like `postcss-calc` to allow for the precision of the decimals to be adjusted.

*note*: this would null my other comment about trailing zeros.

---
Tests I'm envisioning based on my above comments:

`calc(100vw * 1/3 - (30px - 30px * 1/3))` => `calc(100vw * 1/3 - 20px)` **can compress with rounding certainty**
`calc(99.9% * 1/3 - (70px - 70px * 1/3))` => `calc(99.9% * 1/3 - (70px - 70px * 1/3))` **cannot compress with rounding certainty so it's left un-compressed**

**Comment by steve-holland** [comment]
I think some sort of tailboarding process would be great idea. 

I'm starting to think this change (or at least in its current form isn't right). It feels very much as if I'm re-inventing the wheel, as there are already tools out there that simplify the `calc` statement. So I am beginning to wonder if we should just use one of those either as a dependency in our code, or just not do the optimisation and leave other tools like `cssnano` to do the tidy up during the users build.

I think with the rounding we could make it configurable, but then if we start rounding up/down to whole numbers, I've a suspicion we're then going to get issues raised about the columns being 1px out or something like that.

**Comment by peterramsing** [comment]
🤔 I think I agree with not doing the optimization might be the best option. It feels like it could be error prone. 😢

There is a possibility that adding in optimizations that don't require rounding would work but I'm not sure how much of a win that would be.

@codebysubtract Do you think we drop this feature?

**Comment by steve-holland** [comment]
@peterramsing Yes, now the more I think about this, the more I think it could end up doing more harm than good. 

The beauty of Lost is how simple it is. If we start bolting on more and more configuration option, you start moving away from that. Plus, I'm sure most folks already use other plugins like `cssnano`which  help with optimisation so it feels like a duplication of effort anyway.

**Comment by peterramsing** [comment]
@codebysubtract Well, the code looked great. Thanks for the work!

**Pull request closed without merging**
